Tenant-Aware Report filters and ValidatedSW. (#574)
nautobot/nautobot-plugin-device-lifecycle-mgmt: docs/admin/install.md

- `nautobot_lcm_hw_end_of_support_per_part_number`: Number of End of Support devices and inventory items per Part Number.

- `nautobot_lcm_hw_end_of_support_per_location`: Number of End of Support devices and inventory items per Location.

## Multi-Tenant Mode

The `multi_tenant_mode` setting controls which ValidatedSoftware records are considered when evaluating software compliance for a given device.

### Default Behavior (Legacy Mode - `multi_tenant_mode = False`)

**All devices** (whether assigned to a tenant or not) use legacy filtering logic:

- Filter software by **direct device assignment**, **device type**, **device role**, or the device's **tags** (`object_tags`)

> - ValidatedSoftware records with a `device_tenants` assignment are excluded; only untenanted records are considered

Reviewer: If we're going to try to be backward compatible we probably shouldn't change this behavior? In general in Nautobot, assigning a tenant doesn't change any application behavior?

Author: the problem with backwards compatibility was performance (now I remember). A backwards compatible version needs to execute a tenant-mode filter for validated sw and if that does not find a match process the legacy mode filter. This penalizes legacy mode performance.

Reviewer: What I'm suggesting would improve performance. You would remove the `filter(device_tenants__isnull=True)` from `_filter_legacy_mode`

Author: Another reason for that is transition. When a user decides to transition to multi-tenant mode, first it needs to create all the validated software components per tenant. If we modify that filter, during the transition period they would have innacurate software compliance until they flip the switch. This was discussed with @progala (I believe there is even a thread upstream)

- The device's own tenant is ignored — all devices see the same pool of untenanted records

**Use case:** Existing deployments with tenant-assigned devices that don't need per-tenant software isolation.

### Multi-Tenant Mode (`multi_tenant_mode = True`)

Enables **tenant-aware software filtering**:

**For devices WITH a tenant assigned:**

A `ValidatedSoftware` record matches if **any one** of the following applies:

- The device is directly listed in the record's `devices` field (evaluated regardless of whether the record has `device_tenants` set)

> - The record's `device_tenants` includes the device's tenant, AND:

Reviewer: ```suggestion
- The record's `device_tenants` includes the device's tenant, AND one of:
```

    - Both `device_types` and `device_roles` are set and match the device
    - Only `device_types` is set and matches the device's type
    - Only `device_roles` is set and matches the device's role

>     - Neither `device_types` nor `device_roles` is set (tenant-only record)

Reviewer: I don't think tenant should be used as the only criteria for relating a validatedsoftware to a device. Tenant should be an additional filter on top of the existing logic since the tenant model represents administrative ownership: https://docs.nautobot.com/projects/core/en/stable/user-guide/core-data-model/tenancy/tenant/

While tenant could technically be used to match devices to validatedsoftware because both of the models have a relationship to tenant, I think it's incorrect if we don't enforce the real world constraints between the software and the hardware that it is compatible with.

Author: I agree with that. However, allowing tenant only -> validate software assignments has benefits in small multi-tenant deployments, where there is one or a few golden images per platform. In these cases you don't have to create multiple copies of the validated software (device type, role, etc), the user can do one per tenant.

Reviewer: If you have a `ValidatedSoftwareLCM` model with no `device_types`, `device_roles` or `devices` set, how do you know which platform the `ValidatedSoftwareLCM` applies to? Does this fall back to the related `SoftwareVersion` model or does that `ValidatedSoftwareLCM` apply to all devices in that tenant? If that's the case, it only works if that tenant only ever has 1 type of software that it needs to track (Cisco IOS for the 3850 platform for example).

Author: it applies to all devices in the tenant. A good example for this is SDWAN. Clients that segment SDWAN devices using a tenant typically do not want the complexity of creating multiple validated software copies per device type or even role. Just one per tenant.

Reviewer: ValidatedSoftwareLCM has a manytomany field to device types and roles though?

Author: that is correct as well as devices, inventoryitems and tags.

- The record's `object_tags` intersects the device's tags (evaluated regardless of whether the record has `device_tenants` set)

**For devices WITHOUT a tenant assigned:**

> Only `ValidatedSoftware` records with **no `device_tenants` set** are considered.

Reviewer: If the device is explicitly assigned in `ValidatedSoftware.devices` then the tenant match shouldn't matter?

Author: @gsnider2195 this aligns with the concept of "Global" or NO tenant. Basically, in multi-tenant mode we consider no-tenant as a tenant itself.

Reviewer: Can you explicitly state in the docs that when using multi-tenant mode, devices or validatedsoftware with no tenant assigned are considered "Global". That term is used throughout the docs but is never explained.

Author: removed references to `Global` for clarity. user docs (software_lifecycle.md) state: `A non-tenanted device in multi-tenant mode only sees untenanted Validated Software; any record with device_tenants set is filtered out, including via direct assignment and tags.`

A record matches if **any one** of the following applies:

- The device is directly listed in the record's `devices` field
- The record has both `device_types` and `device_roles` set, and both match the device
- The record has only `device_types` set, and it matches the device's type
- The record has only `device_roles` set, and it matches the device's role
- The record's `object_tags` intersects the device's tags

**Use case:** Multi-tenant environments where each tenant has different software requirements and needs isolated software catalogs.

### Configuration

To enable multi-tenant mode, add the following to your `nautobot_config.py`:

```python
PLUGINS_CONFIG = {
    "nautobot_device_lifecycle_mgmt": {
        "multi_tenant_mode": True,  # Enable tenant-aware filtering
    },
}
```

### Examples

#### Example 1: Legacy Mode (Default)

- Device "rtr1" assigned to **Tenant A**

> - Global software "IOS 17.3.3" configured for device type "ASR-1000"

Reviewer: When you say `Global software` are you talking about the `dcim.SoftwareVersion` model or the ValidatedSoftwareLCM model? Since ValidatedSoftwareLCM is typically related to a `dcim.SoftwareVersion`, you should be explicit here to avoid confusion.

Author: good point, making that change.

- `nautobot_lcm_hw_end_of_support_per_part_number`: Number of End of Support devices and inventory items per Part Number.

- `nautobot_lcm_hw_end_of_support_per_location`: Number of End of Support devices and inventory items per Location.

## Multi-Tenant Mode

The `multi_tenant_mode` setting controls how the app filters ValidatedSoftware for devices based on tenant assignments.

### Default Behavior (Legacy Mode - `multi_tenant_mode = False`)

**All devices** (whether assigned to a tenant or not) use legacy filtering logic:
- Filter software by **device type**, **device role**, and **tags only**
- **Tenant assignments are completely ignored**
- Devices in different tenants will see the same software

**Use case:** Existing deployments with tenant-assigned devices that don't need per-tenant software isolation.

### Multi-Tenant Mode (`multi_tenant_mode = True`)

Enables **tenant-aware software filtering**:

**For devices WITH a tenant assigned:**
- Can only see ValidatedSoftware configured for their **specific tenant**

> - Cannot see global/untenanted software

Reviewer 2: This does not seem to be correct. In software_lifecycle.md we say:

---
!!! note
    Tag-based (`object_tags`) and direct device (`devices`) assignments are tenant-agnostic in both modes. A Validated Software record matches a device via these fields regardless of the tenant configuration on the record."

Author: updated

### Configuration

To enable multi-tenant mode, add the following to your `nautobot_config.py`:

```python
PLUGINS_CONFIG = {
    "nautobot_device_lifecycle_mgmt": {
        "multi_tenant_mode": True,  # Enable tenant-aware filtering
    },
}
```

### Examples

#### Example 1: Legacy Mode (Default)

- Device "rtr1" assigned to **Tenant A**
- Global software "IOS 17.3.3" configured for device type "ASR-1000"
- **Result:** Device "rtr1" **CAN** see the global software (legacy behavior)

#### Example 2: Multi-Tenant Mode

- Device "rtr1" assigned to **Tenant A**
- Global software "IOS 17.3.3" configured for device type "ASR-1000" (no tenant)
- ValidatedSoftware for same version configured for **Tenant A**
- **Result:** Device "rtr1" **CANNOT** see the global software, only the Tenant A version

### Switching Modes

To switch between modes, simply update the `multi_tenant_mode` setting in your `nautobot_config.py` and restart Nautobot services:

```shell
sudo systemctl restart nautobot nautobot-worker nautobot-scheduler
```

!!! note
    No database migrations are required when changing this setting - it is purely logical and doesn't affect how software is stored.

> 

Reviewer 2: ```suggestion

!!! warning
    Switching modes in either direction affects compliance results immediately after restart.
    Enabling `multi_tenant_mode` without tenant-scoped Validated Software records will cause
    tenanted devices to show zero matching validations. Disabling it will revert tenanted
    devices to global-only matching; if those devices have no global Validated Software records,
    compliance will also drop to zero until the setting is re-enabled.
```

Author: updated

- `nautobot_lcm_hw_end_of_support_per_part_number`: Number of End of Support devices and inventory items per Part Number.

- `nautobot_lcm_hw_end_of_support_per_location`: Number of End of Support devices and inventory items per Location.

## Multi-Tenant Mode

> The `multi_tenant_mode` setting controls how the app filters ValidatedSoftware for devices based on tenant assignments.

Reviewer 2: ```suggestion
The `multi_tenant_mode` setting controls which ValidatedSoftware records are considered when evaluating software compliance for a given device.
```

### Default Behavior (Legacy Mode - `multi_tenant_mode = False`)

**All devices** (whether assigned to a tenant or not) use legacy filtering logic:

> - Filter software by **direct device assignment**, **device type**, **device role**, and **tags**

Reviewer 2: ```suggestion
- Filter software by **direct device assignment**, **device type**, **device role**, or the device's **tags** (`object_tags`)
```

- ValidatedSoftware records with a `device_tenants` assignment are excluded; only untenanted records are considered
- The device's own tenant is ignored — all devices see the same pool of untenanted records

**Use case:** Existing deployments with tenant-assigned devices that don't need per-tenant software isolation.

### Multi-Tenant Mode (`multi_tenant_mode = True`)

Enables **tenant-aware software filtering**:

**For devices WITH a tenant assigned:**

- See ValidatedSoftware configured for their **specific tenant** (matched by device type, device role, or either being unset)
- Also see ValidatedSoftware matched via **direct device assignment** or **device tags**, regardless of the tenant configured on the record
- Filtering is based on:
    - Direct device assignments (tenant-agnostic)
    - Tenant-specific device type + role matches
    - Tenant-specific device type matches (any role)
    - Tenant-specific device role matches (any type)
    - Tenant-scoped records with no device type or role specified

>     - Device tags (tenant-agnostic)

Reviewer 2: ```suggestion
A `ValidatedSoftware` record matches if **any one** of the following applies:

- The device is directly listed in the record's `devices` field (evaluated regardless of whether the record has `device_tenants` set)
- The record's `device_tenants` includes the device's tenant, AND:
    - Both `device_types` and `device_roles` are set and match the device
    - Only `device_types` is set and matches the device's type
    - Only `device_roles` is set and matches the device's role
    - Neither `device_types` nor `device_roles` is set (tenant-only record)
- The record's `object_tags` intersects the device's tags (evaluated regardless of whether the record has `device_tenants` set)
```

**For devices WITHOUT a tenant assigned:**

- See only ValidatedSoftware records with **no tenant** assigned, matched by direct device assignment, device type, device role, or tags
- Filtering is based on:
    - Direct device assignments (software must have no tenant)
    - Device type and role matches (software must have no tenant)

>     - Device tags (software must have no tenant)

Reviewer 2: ```suggestion
Only `ValidatedSoftware` records with **no `device_tenants` set** are considered.

A record matches if **any one** of the following applies:

- The device is directly listed in the record's `devices` field
- The record has both `device_types` and `device_roles` set, and both match the device
- The record has only `device_types` set, and it matches the device's type
- The record has only `device_roles` set, and it matches the device's role
- The record's `object_tags` intersects the device's tags
```
